### PR TITLE
OCC-195: Boolean product attributes not stored in the order screen.

### LINKED ISSUE #365: Boolean product attributes not stored in the order screen. (OCC-195)
### Describe the bug

The Boolean Product Attribute Field is kind of broken, see symptoms below.

### To Reproduce

1. Edit the "Product" content type (or create a new content type with ProductPart and PricePart).
2. Click _Add Field_ and select _Boolean Product Attribute Field_.
3. Order any Product (e.g. https://localhost:5001/testproduct) without ticking the new checkbox.
4. **In the cart there is no attribute indicator below the product name.**
5. Purchase succeeds, but **the Attributes column is empty** both in Order display and editor.
6. Order the same Product again, but this time check the box.
7. On the cart screen it doesn't show the product
8. Trying to click the Pay button **results in this async error** notification:
```
Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: 'OrchardCore.Commerce.ProductAttributeValues.BooleanProductAttributeValue' does not contain a definition for 'PredefinedValue' at CallSite.Target(Closure , CallSite , Object ) at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0) at OrchardCore.Commerce.Models.ShoppingCartItem.<>c.<CreateOrderLineFromShoppingCartItemAsync>b__22_1(IProductAttributeValue value) in E:\Repositories\orchard-core\OrchardCore.Commerce\src\Modules\OrchardCore.Commerce\Models\ShoppingCartItem.cs:line 127 
```

### Expected behavior

- The attribute should be applied even if its value is false.
- When displaying attributes in the cart or the order view/editor it should display the Boolean attribute in the `{fieldLabel ?? fieldName}: {Yes/No}` format. For example: `Deluxe Version: Yes` or `Insurance: No`
- Including a boolean attribute shouldn't during checkout.
- Boolean attributes should be displayed and editable in the order editor.

If it helps, this attribute type should be considered as if it was predefined with `true` and `false` possible values. It may benefit to have a container value derived from `IPredefinedValuesProductAttributeValue<bool>` similar to `TextProductAttributeValue` for text attributes.

[Jira issue](https://lombiq.atlassian.net/browse/OCC-195)

### PR DESCRIPTION
**Pull request by porgabi**
[OCC-195](https://lombiq.atlassian.net/browse/OCC-195)

Fixes #365

### CODE REVIEW

File: src/Modules/OrchardCore.Commerce/Services/ProductAttributeProvider.cs [hunk 1]
```diff
@@ -28,9 +29,18 @@ public class ProductAttributeProvider : IProductAttributeProvider
...
-            : JsonDocument.Parse(value.ToString()!).RootElement;
+        var element = default(JsonElement);
+        element = value switch
+        {
+            JsonElement jsonElement => jsonElement,
+
+            // String needs to be lowercase for JSON parsing to work.
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            JValue jValue => JsonDocument.Parse(jValue.Value.ToString().ToLowerInvariant()).RootElement,
+#pragma warning restore CA1308 // Normalize strings to uppercase
```
Comment (porgabi) [inline src/Modules/OrchardCore.Commerce/Services/ProductAttributeProvider.cs:40]: Normalizing to uppercase doesn't work in this case:
devenv_LyrfBdXAa3

File: src/Modules/OrchardCore.Commerce/Views/BooleanProductAttributeField.cshtml [hunk 2]
```diff
@@ -7,7 +7,7 @@
     var settings = attr.Settings as BooleanProductAttributeFieldSettings;
     var id = $"{Model.IdPrefix}-attribute-{index}";
 }
-
+<input type="hidden" name="line.Attributes[@index].Value" value="false" />
```
Comment (porgabi) [inline src/Modules/OrchardCore.Commerce/Views/BooleanProductAttributeField.cshtml:10]: Ended up restoring this as without it the attribute does not show up at all due to the missing value.

Comment (porgabi) [reply src/Modules/OrchardCore.Commerce/Views/BooleanProductAttributeField.cshtml:11]: It was causing issues, but then it wasn't and now it isn't. It's weird. UI tests are passing so I suppose it's fine?

File: src/Modules/OrchardCore.Commerce/Models/ShoppingCartItem.cs [hunk 3]
```diff
@@ -124,7 +125,23 @@ public string GetVariantKeyFromAttributes(ISet<string> predefinedAttributeValues
         var price = priceSelectionStrategy.SelectPrice(item.Prices);
         var fullSku = productService.GetOrderFullSku(item, await productService.GetProductAsync(ProductSku));
 
-        var selectedAttributes = Attributes.ToDictionary(key => key.PartName, value => (string)((dynamic)value).PredefinedValue);
+        var selectedTextAttributes = Attributes
+            .Where(attr => attr is TextProductAttributeValue)
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Models/ShoppingCartItem.cs:129]: It's better to use `.CastWhere<T>()` here.

File: src/Modules/OrchardCore.Commerce/Models/ShoppingCartItem.cs [hunk 4]
```diff
@@ -124,7 +125,23 @@ public string GetVariantKeyFromAttributes(ISet<string> predefinedAttributeValues
...
+            .Where(attr => attr is BooleanProductAttributeValue)
+            .ToDictionary(
+                attr => attr.PartName,
+                attr => attr is BooleanProductAttributeValue booleanAttribute ? booleanAttribute.Value.ToString() : null);
+
+        var selectedNumericAttributes = Attributes
+            .Where(attr => attr is NumericProductAttributeValue)
+            .ToDictionary(
+                attr => attr.PartName,
+                attr => attr is NumericProductAttributeValue numericAttribute ? numericAttribute.Value.ToString() : null);
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Models/ShoppingCartItem.cs:144]: This too, should be handled by type specific providers.
Also for the boolean and numeric, you can get the same value with `attr.UntypedValue?.ToString()`.

File: src/Modules/OrchardCore.Commerce/Views/BooleanProductAttributeValue.cshtml [hunk 5]
```diff
@@ -1,12 +1,20 @@
...
-    }
+    var attributeValue = Model.Attribute as BooleanProductAttributeValue;
+    var attributeDescriptions = AttributeService.GetProductAttributeFields(Model.Product as ContentItem);
+    var booleanAttributeFields = attributeDescriptions.Where(attr => attr.Field is BooleanProductAttributeField);
+    var settings = booleanAttributeFields
+        .FirstOrDefault(attr => attr.Name == attributeValue.FieldName).Settings as BooleanProductAttributeFieldSettings;
+
+    TagBuilder tag = Tag(Model, "li");
+    tag.AddCssClass("list-inline-item");
+    tag.InnerHtml.Append((settings.Label ?? attributeValue.FieldName) + ": " + (attributeValue.Value ? "Yes" : "No"));
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Views/BooleanProductAttributeValue.cshtml:18]: Use string localization for the "Yes" and "No".

File: src/Modules/OrchardCore.Commerce/Views/OrderPart_Edit.cshtml [hunk 6]
```diff
@@ -11,18 +12,47 @@
 
 @inject IMoneyService MoneyService
 @inject ISession Session
-@inject IPredefinedValuesProductAttributeService AttributesService
+@inject IPredefinedValuesProductAttributeService PredefinedAttributeService
+@inject IProductAttributeService AttributeService
 
 @{
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Views/OrderPart_Edit.cshtml:18]: You should minimize business logic in cshtml. No static code analysis, error handling works differently, it's not a good time. I think almost the whole of this large code block should be in `OrderPartDisplayDriver.Edit()`

File: test/OrchardCore.Commerce.Tests.UI/Tests/OrderTests/BehaviorOrderTests.cs [hunk 7]
```diff
@@ -137,6 +137,71 @@ public BehaviorOrderTests(ITestOutputHelper testOutputHelper)
                 await context.ClickPublishAsync();
 
                 context.Missing(ByQuantity(1));
+
+                // Boolean attributes should work properly in the Order editor.
+                await context.GoToAdminRelativeUrlAsync("/ContentTypes/AddFieldsTo/Product");
```
Comment (DAud-IcI) [inline test/OrchardCore.Commerce.Tests.UI/Tests/OrderTests/BehaviorOrderTests.cs:142]: The `/ContentTypes/AddFieldsTo/` has been used a few times, please create an extension method for it.

**Review by DAud-IcI — CHANGES_REQUESTED** [review]

File: src/Modules/OrchardCore.Commerce/Drivers/OrderPartDisplayDriver.cs [hunk 8]
```diff
@@ -182,18 +250,51 @@ public override async Task<IDisplayResult> UpdateAsync(OrderPart part, IUpdateMo
...
 
             attributesList.Add(matchingAttribute);
         }
     }
 
+    private (List<IProductAttributeValue> AttributesList,
+        Dictionary<string, string> SelectedTextAttributes,
+        Dictionary<string, string> SelectedBooleanAttributes,
+        Dictionary<string, string> SelectedNumericAttributes)
+        ProcessAttributes(OrderLineItemViewModel lineItem, ProductPart productPart)
```
[the remark this answers is not in the document]
Comment (porgabi) [reply src/Modules/OrchardCore.Commerce/Drivers/OrderPartDisplayDriver.cs:269]: Regarding individual providers, do you mean creating 3 new providers that implement `IProductAttributeProvider` with their own `CreateFromValue()` and `Parse()` implementations?

[the remark this answers is not in the document]
Comment (DAud-IcI) [reply src/Modules/OrchardCore.Commerce/Drivers/OrderPartDisplayDriver.cs:269]: The goal is that we shouldn't have separate variables/properties like `SelectedBooleanAttributes` or `availableBooleanAttributes`. (it's okay to have those in the field-specific views like BooleanProductAttributeField.cshtml and BooleanProductAttributeValue.cshtml)

The provider should interact with dictionaries keyed to the type it handles. So create 3 new providers that implement `IProductAttributeProvider` and wherever you have added type specific logic create a method and move the code to the matching provider. For example the `HandleSelectedBooleanAttributes` should be something like `BooleanProductAttributeProvider.HandleSelected()`.

File: src/Modules/OrchardCore.Commerce/Services/ProductAttributeService.cs [hunk 9]
```diff
@@ -50,6 +50,24 @@ public IEnumerable<ProductAttributeDescription> GetProductAttributeFields(Conten
             .Where(description => description.Field != null);
     }
 
+    public (ContentTypePartDefinition PartDefinition, ContentPartFieldDefinition FieldDefinition)
+        GetFieldDefinition(ContentTypeDefinition type, string attributeName)
+    {
+        var partAndField = attributeName.Split('.');
+        var partName = partAndField[0];
+        var fieldName = partAndField[1];
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Services/ProductAttributeService.cs:58]: We have an extension method for splitting a two-part string by a separator.
```suggestion
        var (partName, _, fieldName) = attributeName.Partition(".");
```

File: src/Modules/OrchardCore.Commerce/Services/StripePaymentService.cs [hunk 10]
```diff
@@ -308,11 +311,22 @@ public async Task<IEnumerable<OrderLineItem>> CreateOrderLineItemsAsync(Shopping
...
 
+            var selectedAttributes = new Dictionary<string, IDictionary<string, string>>();
+            foreach (var provider in _productAttributeProviders)
+            {
+                var attributesByType = provider.GetSelectedAttributes(item.Attributes);
+                if (attributesByType.Any())
+                {
+                    selectedAttributes.Add(attributesByType.Keys.First(), attributesByType.Values.First());
+                }
+            }
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Services/StripePaymentService.cs:322]: This can be converted into linq extensions which would be cleaner.

**Review by DAud-IcI — CHANGES_REQUESTED** [review]

**Comment by DAud-IcI** [comment]
I went ahead and committed the stuff mentioned above, as an apology for the late review.

**Comment by porgabi** [comment]
Thanks! I think it wasn't very late, it's only been a few days, but appreciate it regardless! 😃

**Pull request merged**